allowing to pass a np.random.RandomState to ransac

kevin-keraudren opened the pull request:
## Description
This PR is to allow passing a random_state to `ransac`. Using `np.random.seed()` to freeze the random state is generally a bad idea as it sets a seed for all subsequent `np.random` calls, even outside of `ransac`. For certain reasons, we might want to freeze some seeds while allowing others to be free.

My personal interest in allowing to freeze the random seed of `ransac` are:
 - for unit testing, where `np.random.RandomState` is cleaner than `np.random.seed()` 
 - for running experiments: I have a pipeline where `ransac` is called at the beginning, I want to experiment with different parameters that influence later stages of the pipeline, but I do not want to see in my results the influence of the randomness of ransac (or at least I want to be able to switch between a deterministic `ransac` for a fixed input or a non-deterministic `ransac`).

Additionally, `np.random` has been reported not to be thread-safe, while `np.random.RandomState` is: 
 - https://github.com/scikit-image/scikit-image/issues/1521

## Comments

 - the `check_random_state` function is from scikit-learn: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L564 as well as the docstring line. Please let me know if you would prefer a simpler version that only allows passing a `np.random.RandomState` object. If you think this function is good to use, shall we mention in its docstring that it has been copied from scikit-learn?

 - I updated two of the tests as an example, I can update the remaining ones if you are happy with it.

 - `np.random.RandomState.random` does not exist and `np.random.RandomState.random_sample` should be used instead. Indeed, `np.random.random` is only an alias for `np.random.random_sample` https://github.com/numpy/numpy/blob/master/numpy/random/__init__.py#L102

ahojnnes commented on skimage/measure/fit.py:
```diff
@@ -687,12 +688,29 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
     return int(np.ceil(nom / denom))
+    
+    
+def check_random_state(seed):
```
This is a useful function for other parts of skimage as well. Would you mind moving this to e.g. skimage._shared.util?

ahojnnes commented:
Thanks Kevin, this is a very useful contribution.

ahojnnes commented on skimage/measure/tests/test_fit.py:
```diff
@@ -208,7 +206,7 @@ def test_ransac_shape():
 
     # estimate parameters of corrupted data
-    model_est, inliers = ransac(data0, CircleModel, 3, 5)
+    model_est, inliers = ransac(data0, CircleModel, 3, 5, random_state=np.random.RandomState(1))
```
Move to next line.

ahojnnes commented on skimage/measure/tests/test_fit.py:
```diff
@@ -232,7 +230,7 @@ def test_ransac_geometric():
 
     # estimate parameters of corrupted data
-    model_est, inliers = ransac((src, dst), AffineTransform, 2, 20)
+    model_est, inliers = ransac((src, dst), AffineTransform, 2, 20, random_state=random_state)
```
Here as well, wrap at 80.

kevin-keraudren commented:
@ahojnnes the function `check_random_state()` has been moved to the module `skimage._shared.util`, I added a comment to reference the original function in scikit-learn's module `sklearn.utils.validation`, and I updated `test_fit.py`.

ahojnnes commented on skimage/measure/fit.py:
```diff
@@ -765,6 +765,12 @@ def ransac(data, model_class, min_samples, residual_threshold,
         where the probability (confidence) is typically set to a high value
         such as 0.99, and e is the current fraction of inliers w.r.t. the
         total number of samples.
+    random_state : int, RandomState instance or None, optional (default=None)
```
To comply with our standards, please remove the "(default=None)" part.

ahojnnes commented on skimage/_shared/utils.py:
```diff
@@ -191,3 +192,21 @@ def copy_func(f, name=None):
+
+
+def check_random_state(seed):
+    """Turn seed into a np.random.RandomState instance
```
It would be great if you could transform this into our doc string format. ("Parameters", "Raises", "Returns" section)

ahojnnes commented:
I have two more comments for you, otherwise looks great. Thanks!

kevin-keraudren commented:
@ahojnnes I updated the docstrings, although I do not know what is the convention in scikit-image regarding the use of single or double backticks around variables and module names.

ahojnnes commented on skimage/_shared/utils.py:
```diff
@@ -191,3 +192,30 @@ def copy_func(f, name=None):
+    Parameters
+    ----------
+    seed : None, int or np.random.RandomState
+           If `seed` is None, return the RandomState singleton used by `np.random`.
```
That is following our convention. We use single backticks for variables and double backticks for code.

ahojnnes commented:
This LGTM. Waiting for another core member to agree and merge.

Thanks for your contribution Kevin!

JDWarner commented:
:+1: from me as well. Thanks, @kevin-keraudren!
